<p style="font-family: Arial; font-size:3em;color:black;"> Session 5 - Lab Exercise 10</p>

#### Team Members

1. Kevin Xavier Antony Arul Xavier
2. Vikas Baliyan
3. Ramuel John Batuigas
4. Chirag Ashokkumar Bhatia
5. Jazbi Izhar
6. Sidhartha Sankar Mondal
7. Bharat Sharma

In [2]:
# For this example, we will use K-Means Clustering Project database from Kaggle (https://www.kaggle.com/faressayah/k-means-clustering-private-vs-public-universities)
# We actually have the labels for this data set, but we will NOT use them for the KMeans clustering algorithm, since that is an unsupervised learning algorithm.
# As we will shortly see, we have a data frame with 777 observations on 18 variables.

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [24]:
df = pd.read_csv('College_Data.csv',index_col=0)
df.columns

Index(['Private', 'Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc',
       'F.Undergrad', 'P.Undergrad', 'Outstate', 'Room.Board', 'Books',
       'Personal', 'PhD', 'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend',
       'Grad.Rate'],
      dtype='object')

In [25]:
df['Grad.Rate']['Cazenovia College'] = 100

# Try removing various columns (features) from the dataset and examin if it improves/degrades your K-Means model performance, or it may have little impact.
# Report 10 cases where you removed one or more features and indicate how it impacted the model performance.

/var/folders/zq/mhspsl9536b545q5z960zpnm0000gn/T/ipykernel_17877/3450590736.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Grad.Rate']['Cazenovia College'] = 100


In [38]:
df['Private'] = df['Private'].replace({'Yes':1, 'No':0})
y = df['Private']
df1 = df.drop('Private',axis=1)

In [27]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import accuracy_score

In [28]:
def KNN_accuracy(X,y):
    print(X.columns)
    X_normalized = MinMaxScaler().fit_transform(X)
    model = KMeans(n_clusters=2)
    model.fit(X_normalized)
    pred = model.labels_
    return accuracy_score(y,pred)

In [31]:
acc1 = KNN_accuracy(df1,y)
acc1

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD',
       'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.5495495495495496

In [32]:

acc2 = KNN_accuracy(df1.drop(['Books'],axis=1),y)
acc2

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Personal', 'PhD', 'Terminal',
       'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.45302445302445304

In [36]:
acc3 = KNN_accuracy(df1.drop(['PhD'],axis=1),y)
acc3

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal',
       'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.4144144144144144

In [40]:
acc4 = KNN_accuracy(df1.drop(['Top25perc'],axis=1),y)
acc4

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'F.Undergrad', 'P.Undergrad',
       'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD', 'Terminal',
       'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.6061776061776062

In [41]:
acc5 = KNN_accuracy(df1.drop(['Top10perc'],axis=1),y)
acc5

Index(['Apps', 'Accept', 'Enroll', 'Top25perc', 'F.Undergrad', 'P.Undergrad',
       'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD', 'Terminal',
       'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.41312741312741313

In [42]:
acc6 = KNN_accuracy(df1.drop(['F.Undergrad' , 'P.Undergrad'],axis=1),y)
acc6

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'Outstate',
       'Room.Board', 'Books', 'Personal', 'PhD', 'Terminal', 'S.F.Ratio',
       'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.44787644787644787

In [43]:
acc7 = KNN_accuracy(df1.drop(['Outstate' , 'Room.Board'],axis=1),y)
acc7

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Books', 'Personal', 'PhD', 'Terminal', 'S.F.Ratio',
       'perc.alumni', 'Expend', 'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.4967824967824968

In [46]:
acc8 = KNN_accuracy(df1.drop(['PhD','Terminal',	'S.F.Ratio', 'perc.alumni'],axis=1),y)
acc8

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'Expend',
       'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.46589446589446587

In [44]:
acc9 = KNN_accuracy(df1.drop(['Top10perc', 'Top25perc', 'F.Undergrad' , 'P.Undergrad'],axis=1),y)
acc9

Index(['Apps', 'Accept', 'Enroll', 'Outstate', 'Room.Board', 'Books',
       'Personal', 'PhD', 'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend',
       'Grad.Rate'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.6254826254826255

In [47]:
acc10 = KNN_accuracy(df1.drop(['Grad.Rate' , 'Terminal'],axis=1),y)
acc10

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD',
       'S.F.Ratio', 'perc.alumni', 'Expend'],
      dtype='object')


/Users/Shared/miniconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


0.5328185328185329

### Answer:

From the cases demonstrated above, we can see that dropping some columns in the dataframe affect the accuracy score of KNN classification. For the cases above, the highest accuracy score was achieved after dropping columns 'Top10perc', 'Top25perc', 'F.Undergrad' , 'P.Undergrad' resulting in an accuracy score of 0.62. The lowest accuracy score was taken from the case where droppping 'Top10perc' column gives out 0.41 accuracy.